src/Constraint/Constraint.php

```diff
@@ -128,13 +151,18 @@ public function versionCompare($a, $b, $operator, $compareBranches = false)
     }
 
     /**
-     * @param Constraint $provider
+     * @param ConstraintInterface $provider
      * @param bool $compareBranches
      *
      * @return bool
      */
-    public function matchSpecific(Constraint $provider, $compareBranches = false)
+    public function matches(ConstraintInterface $provider, $compareBranches = false)
     {
+        if ($provider instanceof MultiConstraint) {
+            // turn matching around to find a match
+            return $provider->matches($this);
+        }
```

Comment:
This means that an EmptyConstraint should match anything, which will not work anymore with your new code.

I suggest changing the check from ``if ($provider instanceof MultiConstraint) {`` to ``if (!$provider instanceof self) {``